core/block_validator.go

```diff
@@ -65,56 +66,81 @@ func (v *BlockValidator) ValidateBody(block *types.Block) error {
+				if block.Withdrawals() == nil {
+					return errors.New("missing withdrawals in block body")
+				}
+				if hash := types.DeriveSha(block.Withdrawals(), trie.NewStackTrie(nil)); hash != *header.WithdrawalsHash {
+					return fmt.Errorf("withdrawals root hash mismatch (header value %x, calculated %x)", *header.WithdrawalsHash, hash)
+				}
+			} else if block.Withdrawals() != nil {
+				// Withdrawals are not allowed prior to Shanghai fork
+				return errors.New("withdrawals present in block body")
+			}
+			return nil
+		},
+		func() error {
+			// Blob transactions may be present after the Cancun fork.
+			var blobs int
+			for i, tx := range block.Transactions() {
+				// Count the number of blobs to validate against the header's blobGasUsed
+				blobs += len(tx.BlobHashes())
 
-		// The individual checks for blob validity (version-check + not empty)
-		// happens in StateTransition.
-	}
+				// If the tx is a blob tx, it must NOT have a sidecar attached to be valid in a block.
+				if tx.BlobTxSidecar() != nil {
+					return fmt.Errorf("unexpected blob sidecar in transaction at index %d", i)
+				}
 
-	// Check blob gas usage.
-	if header.BlobGasUsed != nil {
-		if want := *header.BlobGasUsed / params.BlobTxBlobGasPerBlob; uint64(blobs) != want { // div because the header is surely good vs the body might be bloated
-			return fmt.Errorf("blob gas used mismatch (header %v, calculated %v)", *header.BlobGasUsed, blobs*params.BlobTxBlobGasPerBlob)
-		}
-	} else {
-		if blobs > 0 {
-			return errors.New("data blobs present in block body")
-		}
-	}
+				// The individual checks for blob validity (version-check + not empty)
+				// happens in StateTransition.
+			}
 
-	// Ancestor block must be known.
-	if !v.bc.HasBlockAndState(block.ParentHash(), block.NumberU64()-1) {
-		if !v.bc.HasBlock(block.ParentHash(), block.NumberU64()-1) {
-			return consensus.ErrUnknownAncestor
+			// Check blob gas usage.
+			if header.BlobGasUsed != nil {
+				if want := *header.BlobGasUsed / params.BlobTxBlobGasPerBlob; uint64(blobs) != want { // div because the header is surely good vs the body might be bloated
+					return fmt.Errorf("blob gas used mismatch (header %v, calculated %v)", *header.BlobGasUsed, blobs*params.BlobTxBlobGasPerBlob)
+				}
+			} else {
+				if blobs > 0 {
+					return errors.New("data blobs present in block body")
+				}
+			}
+			return nil
+		},
+		func() error {
+			// Ancestor block must be known.
+			if !v.bc.HasBlockAndState(block.ParentHash(), block.NumberU64()-1) {
+				if !v.bc.HasBlock(block.ParentHash(), block.NumberU64()-1) {
+					return consensus.ErrUnknownAncestor
+				}
+				return consensus.ErrPrunedAncestor
+			}
+			return nil
+		},
+	}
+	validateRes := make(chan error, len(validateFuns))
+	for _, f := range validateFuns {
+		tmpFunc := f
+		gopool.Submit(func() {
+			validateRes <- tmpFunc()
+		})
+	}
+	for i := 0; i < len(validateFuns); i++ {
+		err := <-validateRes
+		if err != nil {
+			return err
```

Comment:
Hello, in versions before v0.3.0, the op-geth code used to have similar parallel logic here. Our audit partner pointed out the instability of this logic: in the previous serial logic, the function only had a chance to return consensus.ErrUnknownAncestor error or consensus.ErrPrunedAncestor error if all other validations passed. However, now if consensus.ErrPrunedAncestor is returned first, even if other processes return regular errors, the entire function will return consensus.ErrPrunedAncestor. And consensus.ErrPrunedAncestor will change the logic branch of the outer code.
Therefore, according to the audit requirements, I modified the code at the time to fill this vulnerability, please refer to: https://github.com/bnb-chain/op-geth/pull/14
Please modify it according to the PR and correct the logic.

Reply:
Fixed.